ShowOnlyCurrentBranch were not showing commits

gerhardol opened the pull request:
Fixes #10225

## Proposed changes

Git Notes must be excluded for --all only with the previous argument "--not --glob=notes --not".
Changed to the easier to understand "--exclude=refs/notes" that could be used independent of the various branch filters (including reflog).
However, "--exclude=refs/notes" only makes sense with --all, so kept there.

## Test methodology <!-- How did you ensure quality? -->

Adopted test.
(It is really of the kind to increase test coverage, so quite meaningless.)

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
@mstv should we show stash and notes for current only and branches?
Not done in 3.5

RussKie reviewed (commented):
I've looked at the changes but I can't currently verify the correctness of the changes.

I think we ought to come up with integration tests which would exercise various filters and assert mber of entries in the revision grid.

gerhardol commented:
> I've looked at the changes but I can't currently verify the correctness of the changes.

You will need to read https://git-scm.com/docs/git-log and understand the order and dependency of options.
The handling was quite complicated before when the implementation was split in three modules, should be easier now.

> I think we ought to come up with integration tests which would exercise various filters and assert mber of entries in the revision grid.

There are tests, some added in the recent refactoring. What was missing here was that the option was dependent  for this option (most options are unrelated). That test was added (will be added for stash too). The option was also changed to be easier to understand from the man page.

I have not used this myself yet, it is therefore this is a draft. (Was too late when creating.)

gerhardol commented:
CodeFactor reports an issue due to the added test case generator FilterInfo_NotesStash

mstv commented:
> Should we show stash and notes for current only and branches?

I think at least `stash@{0}` should always be displayed. Stashes are one's own work. (The `Show stashes` option could be added to the Advanced Filter - only if someone complains.)

I don't use notes (and I am not fond of auxilary commits). Though if one uses them, these commits could be of interest - at least if they are related to the visible branches.

mstv commented on GitUI/UserControls/RevisionGrid/FilterInfo.cs:
```diff
@@ -308,44 +308,42 @@ public ArgumentString GetRevisionFilter()
-                {
-                    // All branches
-                    filter.Add("--all");
+                // All refs/
+                filter.Add("--all");
```
❔ Adding this argument in front of the "--exclude"s would be clearer to me. (I have compared this with the `git log` doc yet.)

gerhardol commented:
> I think at least `stash@{0}` should always be displayed. Stashes are one's own work. (The `Show stashes` option could be added to the Advanced Filter - only if someone complains.)

Added, filter is weird though...

> I don't use notes (and I am not fond of auxilary commits). Though if one uses them, these commits could be of interest - at least if they are related to the visible branches.

These commits could be excluded also for --all, not of interest. The commit they relate are not seen.
![image](https://user-images.githubusercontent.com/6248932/194427581-9fe04169-da12-4b76-9ffb-f8f04c1afc85.png)

gerhardol commented:
Changed the filter to include stash@{0}, added a test.
I was tempted to make this ref independent of the branch filter, but it takes about 1 ms to resolve this ref for me, so it is dependent of current/filtered branches

--
I have 563 stashes in my GE test repo and to resolve all of them 100 times with --all, limiting output to 100 commits and that requires about 1:38 instead of 48 s, which indicates one stash is resolved in 1 ms. As the tests are repeated, a real scenario without caching could require more time. 
So always adding `stash@{0}` will add 1 ms to --reflog and -all, even if not needed. But the tests and the flow would be simpler.

mstv commented:
The test `GitExtensions.UITests.UserControls.RevisionGrid.RevisionGridControlTests.View_reflects_reset_branch_filter...` seems to fail consistently.

`stash@{0}` makes git think this was the HEAD of the `Current branch (only)`.

I think we must create a (set of) test repo(s) and verify that the expected commits are returned by git.

gerhardol commented:
> The test `GitExtensions.UITests.UserControls.RevisionGrid.RevisionGridControlTests.View_reflects_reset_branch_filter...` seems to fail consistently.
> 
> `stash@{0}` makes git think this was the HEAD of the `Current branch (only)`.
> 
> I think we must create a (set of) test repo(s) and verify that the expected commits are returned by git.

I pushed an update. If there were no stashes, the query failed. So back to a convoluted logic to get stash@{0} in the grid for current only...

mstv commented:
When refreshing (or switching between `All branches` and `Filtered branches` with empty filter), the result varies randomly:

![image](https://user-images.githubusercontent.com/36601201/194653789-ea9d3c9a-83b3-47cd-bd96-1907f17916d8.png)

![image](https://user-images.githubusercontent.com/36601201/194653990-bcdcdcf5-27fc-44a7-a394-0af0ab6b08f9.png)

![image](https://user-images.githubusercontent.com/36601201/194654304-2c8ca1fa-390a-4537-9fd3-0a79f93807b6.png)

![image](https://user-images.githubusercontent.com/36601201/194654627-c3bd8559-6615-4b92-9344-f09972c11d86.png)

![image](https://user-images.githubusercontent.com/36601201/194654192-d340e291-37f5-495d-bb79-03f81b769446.png)

gerhardol commented:
> When refreshing (or switching between `All branches` and `Filtered branches` with empty filter), 

That is the same...

> the result varies randomly:

The Git arguments is the same in this scenario, so it is a different issue
I cannot see why the revisions are processed in a different way.

gerhardol commented:
Squashed the changes for "stash@{0}" for current/filtered. That commit could be separated too (I prefer to keep it in this though).

mstv reviewed (approved):
Thank you for working this out

gerhardol commented:
Rebase merge this tomorrow
